flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/OutputColsHelper.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.common.utils;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.types.Row;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+
+/**
+ * Util for generating output schema when doing prediction or transformation.
+ *
+ * <p>Input:
+ * 1) Schema of input data being predicted or transformed.
+ * 2) Output column names of the prediction/transformation operator.
+ * 3) Output column types of the prediction/transformation operator.
+ * 4) Reserved column names, which is a subset of input data's column names that we want to preserve.
+ *
+ * <p>Output:
+ * 1）The result data schema. The result data is a combination of the preserved columns and the operator's
+ * output columns.
+ *
+ * <p>Several rules are followed:
+ * 1) If reserved columns are not given, then all columns of input data is reserved.
+ * 2）The reserved columns are arranged ahead of the operator's output columns in the final output.
+ * 3) If some of the reserved column names overlap with those of operator's output columns, then the operator's
+ * output columns override the conflicting reserved columns.
+ *
+ * <p>For example, if we have input data schema of ["id":INT, "f1":FLOAT, "f2":DOUBLE], and the operator outputs
+ * a column "label" with type STRING, and we want to preserve the column "id", then we get the output
+ * schema of ["id":INT, "label":STRING].
+ */
+public class OutputColsHelper implements Serializable {
```

Comment:
based on the usage #9413 and #9523 . this can be private to flink, yes? (I do not foresee user directly interacting with this Helper function)

Reply:
Yes, this helper function is for algo developers, not for users.

Reply:
I would suggest putting this as part of the Javadoc
--> Unlike Scala Java has no module private modifier like `private[flink]`, but it is still good to point out IMO

Reply:
In response to my previous comment. I would add a paragraph to make clear in the Javadoc that: end user should not directly interact with this helper class. instead it will be indirectly used via concrete algorithms.